tracing/tracing.go

```diff
@@ -0,0 +1,187 @@
+		m.shutdownFunc = nil
+		otel.SetTracerProvider(trace.NewNoopTracerProvider())
+		level.Info(m.logger).Log("msg", "Tracing provider uninstalled.")
+		return nil
+	}
+
+	tp, shutdownFunc, err := buildTracerProvider(context.Background(), cfg.TracingConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to install a new tracer provider")
+	}
+
+	m.shutdownFunc = shutdownFunc
+	m.config = cfg.TracingConfig
+	otel.SetTracerProvider(tp)
+
+	level.Info(m.logger).Log("msg", "Successfully installed a new tracer provider.")
+	return nil
+}
+
+// Stop gracefully shuts down the tracer provider and stops the tracing manager.
+func (m *Manager) Stop() {
+	defer close(m.done)
+
+	if m.shutdownFunc == nil {
+		return
+	}
+
+	if err := m.shutdownFunc(); err != nil {
+		level.Error(m.logger).Log("msg", "failed to shut down the tracer provider", "err", err)
+	}
+
+	level.Info(m.logger).Log("msg", "Tracing manager stopped")
+}
+
+type otelErrHandler func(err error)
+
+func (o otelErrHandler) Handle(err error) {
+	o(err)
+}
+
+// buildTracerProvider return a new tracer provider ready for installation, together
+// with a shutdown function.
+func buildTracerProvider(ctx context.Context, tracingCfg config.TracingConfig) (trace.TracerProvider, func() error, error) {
+	exp, err := otlptrace.New(ctx, getClient(tracingCfg))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a resource describing the service and the runtime.
+	res, err := resource.New(
+		ctx,
+		resource.WithSchemaURL(semconv.SchemaURL),
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String(serviceName),
+			semconv.ServiceVersionKey.String(version.Version),
+		),
+		resource.WithProcessRuntimeDescription(),
+		resource.WithTelemetrySDK(),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exp),
+		tracesdk.WithSampler(tracesdk.ParentBased(
+			tracesdk.TraceIDRatioBased(tracingCfg.SamplingFraction),
+		)),
+		tracesdk.WithResource(res),
+	)
+
+	return tp, func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		err := tp.Shutdown(ctx)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}, nil
+}
+
+// getClient returns an appropriate OTLP client (either gRPC or HTTP), based
+// on the provided tracing configuration.
+func getClient(tracingCfg config.TracingConfig) otlptrace.Client {
+	var client otlptrace.Client
+	switch tracingCfg.ClientType {
+	case config.TracingClientGRPC:
+		opts := []otlptracegrpc.Option{otlptracegrpc.WithEndpoint(tracingCfg.Endpoint)}
+		if !tracingCfg.WithSecure {
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+
+		client = otlptracegrpc.NewClient(opts...)
+	case config.TracingClientHTTP:
+		opts := []otlptracehttp.Option{otlptracehttp.WithEndpoint(tracingCfg.Endpoint)}
+		if !tracingCfg.WithSecure {
+			opts = append(opts, otlptracehttp.WithInsecure())
+		}
+		client = otlptracehttp.NewClient(opts...)
+	}
+
+	return client
```

Comment:
There's a difference between each case, since once we're calling `NewClient` for gRPC (from `otlptracegrpc` package), the other time it's called from `otlptracehttp` for HTTP client. 

With this change, we would always be creating HTTP client from `otlptracehttp.NewClient`.

Reply:
Ah didn't notice that, thanks.